src/api/app/models/workflow/step/set_flags.rb

```diff
@@ -0,0 +1,95 @@
+class Workflow::Step::SetFlags < ::Workflow::Step
+  REQUIRED_KEYS = [:flags].freeze
+  REQUIRED_FLAG_KEYS = [:type, :status, :project].freeze
+  OPTIONAL_FLAG_KEYS = [:package, :repository, :architecture].freeze
+
+  validate :validate_project_and_package
+  validate :validate_flags
+
+  def call(_options = {})
+    return unless valid?
+
+    set_flags
+  end
+
+  def flags
+    step_instructions[:flags]
+  end
+
+  private
+
+  def set_flags
+    ActiveRecord::Base.transaction do
+      flags.each do |flag|
+        project = Project.find_by(name: target_project_name(project_name: flag[:project]))
+        package = project.packages.find_by(name: target_package_name(package_name: flag[:package])) if project.present? && flag[:package].present?
+        main_object = package.presence || project
+        check_access(main_object)
+
+        architecture_id = Architecture.find_by_name(flag[:architecture]).id if flag[:architecture]
+        existing_flag = main_object.flags.exists?(flag: flag[:type], repo: flag[:repository], architecture_id: architecture_id)
+        next if existing_flag
+
+        main_object.add_flag(flag[:type], flag[:status], flag[:repository], flag[:architecture])
+        main_object.save!
+      end
+    end
+  end
+
+  def check_access(object)
+    raise Pundit::NotAuthorizedError unless Pundit.policy(token.user, object).update?
+  end
+
+  def validate_project_and_package
```

Comment:
@rubhanazeem and I talked about this validation and it won't be possible to validate the existence of projects/packages. `set_flags` depends on records (projects/packages/repositories) created by previous steps in a workflow, then applies flags to those records. Since we validate **all** steps before running them, we cannot ensure the existence of those records, so this validation cannot reliably work. Validating step one by one is not possible without having to potentially run only parts of a workflow if certain steps are non-valid. This can lead to major issues and this is why we started to validate **all** steps before running them.